### PR TITLE
wasm: Support WASM based in_exec_wasi and filter_wasm plugins

### PR DESCRIPTION
**Pull request by cosmo0920**
In Fluent Bit 2.0 release, we plan to extend WASM handling on Input and Filter plugins.

This work's motivations are:

1. Open doors the possibility of plugin developments to the C with WASI SDK, TinyGO which is the subset of Golang, and Rust.
2. Execute with custom procedure with sandboxed environment. 
    * Development with C sometimes causes SEGV. This is not welcomed for users. 
    * User-developed plugins also not well-behaved than the core plugins. This is one of the obstacles for extensibility.

 
For handling WASM on input plugin, we plan to support WASI format of WASM program that is self-consistent and having `_start` entrypoint. Ingesting logs with stdout.

For handing WASM on filter plugin, we plan to support WASI format of WASM program that uses C ABI.
TinyGO always uses C ABI to interact for C runtime. And for Rust, we'll use `wasm32-unknown-unknown` ABI instead of `wasm32-wasi` ABI that is still living Rust's [String](https://doc.rust-lang.org/std/string/struct.String.html) and [str](https://doc.rust-lang.org/std/primitive.str.html) types. They have different behavior for `*const c_char` on Rust's FFI.

### Plugin components

* in_exec_wasi: This plugin handles WASI format of WASM like as executables. It communicates with WASI via stdout.
* filter_wasm: This plugin handles WASI format of WASM like as filters. It communicates with WASI via buffer copying from host buffers.

Both of plugins are built on the top of `flb_wasm` abstraction layer. This is because wasm-micro-runtime (WAMR) sometimes breaks the normal definitions of C types. So, we have to isolate and create abstraction layer for WAMR.

### Components dependencies especially for WASM

Note that flb-wamrc is adding `flb-` prefixed executable to provide AOT compiler from wasm-micro-runtime. This AOT compiler will be statically linking LLVM related libraries.
Currently, building `flb-wamrc` is disabled by default.

```mermaid
graph TD;
    in_exec_wasi-->flb_wasm;
    filter_wasm-->flb_wasm;
    flb_wasm-->wasm-micro-runtime;
    flb-wamrc-->wasm-micro-runtime;
    flb-wamrc-->LLVM;
```

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change

### in_exec_wasi

```ini
[SERVICE]
    Flush        1
    Daemon       Off
    Parsers_File parsers.conf
    Log_Level    info
    HTTP_Server  Off
    HTTP_Listen  0.0.0.0
    HTTP_Port    2020

[INPUT]
    Name exec_wasi
    Tag  exec.wasi.local
    WASI_Path wasi_serde_json.wasm
    Parser wasi
    accessible_paths .,/media/Data2/Gitrepo/fluent-bit

[OUTPUT]
    Name  stdout
    Match
```

### filter_wasm

```ini
[SERVICE]
    Flush        1
    Daemon       Off
    Log_Level    info
    HTTP_Server  Off
    HTTP_Listen  0.0.0.0
    HTTP_Port    2020

[INPUT]
    Name dummy
    Tag dummy.locals

[FILTER]
    Name wasm
    match dummy.*
    WASM_Path filter_rust.wasm
    Function_Name rust_filter
    accessible_paths .,/media/Data2/Gitrepo/fluent-bit

[OUTPUT]
    Name  stdout
    Match *
```
- [x] Debug log output from testing the change

### in_exec_wasi

```log
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/08/25 10:24:57] [ info] Configuration:
[2022/08/25 10:24:57] [ info]  flush time     | 1.000000 seconds
[2022/08/25 10:24:57] [ info]  grace          | 5 seconds
[2022/08/25 10:24:57] [ info]  daemon         | 0
[2022/08/25 10:24:57] [ info] ___________
[2022/08/25 10:24:57] [ info]  inputs:
[2022/08/25 10:24:57] [ info]      exec_wasi
[2022/08/25 10:24:57] [ info] ___________
[2022/08/25 10:24:57] [ info]  filters:
[2022/08/25 10:24:57] [ info] ___________
[2022/08/25 10:24:57] [ info]  outputs:
[2022/08/25 10:24:57] [ info]      stdout.0
[2022/08/25 10:24:57] [ info] ___________
[2022/08/25 10:24:57] [ info]  collectors:
[2022/08/25 10:24:57] [ info] [fluent bit] version=2.0.0, commit=444364edba, pid=1177541
[2022/08/25 10:24:57] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/08/25 10:24:57] [ info] [storage] version=1.2.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/08/25 10:24:57] [ info] [cmetrics] version=0.3.5
[2022/08/25 10:24:57] [debug] [exec_wasi:exec_wasi.0] created event channels: read=21 write=22
[2022/08/25 10:24:57] [debug] [input:exec_wasi:exec_wasi.0] interval_sec=1 interval_nsec=0 oneshot=0 buf_size=4096
[2022/08/25 10:24:57] [debug] [stdout:stdout.0] created event channels: read=23 write=24
[2022/08/25 10:24:57] [debug] [router] match rule exec_wasi.0:stdout.0
[2022/08/25 10:24:57] [ info] [sp] stream processor started
[2022/08/25 10:24:57] [ info] [output:stdout:stdout.0] worker #0 started
[2022/08/25 10:24:58] [debug] [input chunk] update output instances with new chunk size diff=63
[2022/08/25 10:24:59] [debug] [task] created task=0x7fa965692960 id=0 OK
[2022/08/25 10:24:59] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] exec.wasi.local: [1661390698.580142345, {"age"=>43, "name"=>"John Doe", "phones"=>["+44 1234567", "+44 2345678"]}]
[2022/08/25 10:24:59] [debug] [out flush] cb_destroy coro_id=0
[2022/08/25 10:24:59] [debug] [input chunk] update output instances with new chunk size diff=63
[2022/08/25 10:24:59] [debug] [task] destroy task=0x7fa965692960 (task_id=0)
[2022/08/25 10:25:00] [debug] [task] created task=0x7fa965692960 id=0 OK
[2022/08/25 10:25:00] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] exec.wasi.local: [1661390699.577202323, {"age"=>43, "name"=>"John Doe", "phones"=>["+44 1234567", "+44 2345678"]}]
[2022/08/25 10:25:00] [debug] [out flush] cb_destroy coro_id=1
[2022/08/25 10:25:00] [debug] [input chunk] update output instances with new chunk size diff=63
[2022/08/25 10:25:00] [debug] [task] destroy task=0x7fa965692960 (task_id=0)
^C[2022/08/25 10:25:00] [engine] caught signal (SIGINT)
[2022/08/25 10:25:00] [debug] [task] created task=0x7fa965692960 id=0 OK
[2022/08/25 10:25:00] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/08/25 10:25:00] [ warn] [engine] service will shutdown in max 5 seconds
[0] exec.wasi.local: [1661390700.576773214, {"age"=>43, "name"=>"John Doe", "phones"=>["+44 1234567", "+44 2345678"]}]
[2022/08/25 10:25:00] [debug] [out flush] cb_destroy coro_id=2
[2022/08/25 10:25:00] [debug] [task] destroy task=0x7fa965692960 (task_id=0)
[2022/08/25 10:25:01] [ info] [engine] service has stopped (0 pending tasks)
[2022/08/25 10:25:01] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/08/25 10:25:01] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

### filter_wasm

```log
Fluent Bit v2.0.0
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/08/25 09:50:38] [ info] Configuration:
[2022/08/25 09:50:38] [ info]  flush time     | 1.000000 seconds
[2022/08/25 09:50:38] [ info]  grace          | 5 seconds
[2022/08/25 09:50:38] [ info]  daemon         | 0
[2022/08/25 09:50:38] [ info] ___________
[2022/08/25 09:50:38] [ info]  inputs:
[2022/08/25 09:50:38] [ info]      dummy
[2022/08/25 09:50:38] [ info] ___________
[2022/08/25 09:50:38] [ info]  filters:
[2022/08/25 09:50:38] [ info]      wasm.0
[2022/08/25 09:50:38] [ info] ___________
[2022/08/25 09:50:38] [ info]  outputs:
[2022/08/25 09:50:38] [ info]      stdout.0
[2022/08/25 09:50:38] [ info] ___________
[2022/08/25 09:50:38] [ info]  collectors:
[2022/08/25 09:50:38] [ info] [fluent bit] version=2.0.0, commit=444364edba, pid=1171979
[2022/08/25 09:50:38] [debug] [engine] coroutine stack size: 24576 bytes (24.0K)
[2022/08/25 09:50:38] [ info] [storage] version=1.2.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
[2022/08/25 09:50:38] [ info] [cmetrics] version=0.3.5
[2022/08/25 09:50:38] [debug] [dummy:dummy.0] created event channels: read=21 write=22
[2022/08/25 09:50:38] [debug] [stdout:stdout.0] created event channels: read=23 write=24
[2022/08/25 09:50:38] [debug] [router] match rule dummy.0:stdout.0
[2022/08/25 09:50:38] [ info] [sp] stream processor started
[2022/08/25 09:50:38] [ info] [output:stdout:stdout.0] worker #0 started
[2022/08/25 09:50:39] [debug] [input chunk] update output instances with new chunk size diff=124
[2022/08/25 09:50:40] [debug] [task] created task=0x7f0d20c42540 id=0 OK
[2022/08/25 09:50:40] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.locals: [1661388639.053042135, {"lang"=>"Rust", "message"=>"dummy", "original"=>"{"message":"dummy"}", "tag"=>"dummy.locals", "time"=>"2022-08-25T00:50:39.053042135 +0000"}]
[2022/08/25 09:50:40] [debug] [out flush] cb_destroy coro_id=0
[2022/08/25 09:50:40] [debug] [input chunk] update output instances with new chunk size diff=124
[2022/08/25 09:50:40] [debug] [task] destroy task=0x7f0d20c42540 (task_id=0)
[2022/08/25 09:50:41] [debug] [task] created task=0x7f0d20c42540 id=0 OK
[2022/08/25 09:50:41] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.locals: [1661388640.053021750, {"lang"=>"Rust", "message"=>"dummy", "original"=>"{"message":"dummy"}", "tag"=>"dummy.locals", "time"=>"2022-08-25T00:50:40.053021750 +0000"}]
[2022/08/25 09:50:41] [debug] [out flush] cb_destroy coro_id=1
[2022/08/25 09:50:41] [debug] [input chunk] update output instances with new chunk size diff=124
[2022/08/25 09:50:41] [debug] [task] destroy task=0x7f0d20c42540 (task_id=0)
[2022/08/25 09:50:42] [debug] [task] created task=0x7f0d20c42540 id=0 OK
[2022/08/25 09:50:42] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[0] dummy.locals: [1661388641.053043212, {"lang"=>"Rust", "message"=>"dummy", "original"=>"{"message":"dummy"}", "tag"=>"dummy.locals", "time"=>"2022-08-25T00:50:41.053043212 +0000"}]
[2022/08/25 09:50:42] [debug] [out flush] cb_destroy coro_id=2
[2022/08/25 09:50:42] [debug] [input chunk] update output instances with new chunk size diff=124
[2022/08/25 09:50:42] [debug] [task] destroy task=0x7f0d20c42540 (task_id=0)
^C[2022/08/25 09:50:42] [engine] caught signal (SIGINT)
[0] dummy.locals: [1661388642.053046150, {"lang"=>"Rust", "message"=>"dummy", "original"=>"{"message":"dummy"}", "tag"=>"dummy.locals", "time"=>"2022-08-25T00:50:42.053046150 +0000"}]
[2022/08/25 09:50:42] [debug] [task] created task=0x7f0d20c42540 id=0 OK
[2022/08/25 09:50:42] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2022/08/25 09:50:42] [debug] [out flush] cb_destroy coro_id=3
[2022/08/25 09:50:42] [ warn] [engine] service will shutdown in max 5 seconds
[2022/08/25 09:50:42] [ info] [input] pausing dummy.0
[2022/08/25 09:50:42] [debug] [task] destroy task=0x7f0d20c42540 (task_id=0)
[2022/08/25 09:50:43] [ info] [engine] service has stopped (0 pending tasks)
[2022/08/25 09:50:43] [ info] [input] pausing dummy.0
[2022/08/25 09:50:43] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/08/25 09:50:43] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

### in_exec_wasi

```log
==1177797== 
==1177797== HEAP SUMMARY:
==1177797==     in use at exit: 0 bytes in 0 blocks
==1177797==   total heap usage: 26,302 allocs, 26,302 frees, 10,857,283 bytes allocated
==1177797== 
==1177797== All heap blocks were freed -- no leaks are possible
==1177797== 
==1177797== Use --track-origins=yes to see where uninitialised values come from
==1177797== For lists of detected and suppressed errors, rerun with: -s
==1177797== ERROR SUMMARY: 11 errors from 9 contexts (suppressed: 0 from 0)
```

### filter_wasm

```log
==1172056== 
==1172056== HEAP SUMMARY:
==1172056==     in use at exit: 0 bytes in 0 blocks
==1172056==   total heap usage: 1,426 allocs, 1,426 frees, 1,759,641 bytes allocated
==1172056== 
==1172056== All heap blocks were freed -- no leaks are possible
==1172056== 
==1172056== For lists of detected and suppressed errors, rerun with: -s
==1172056== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [ ] Attached [local packaging test](./packaging/local-build-all.sh) output showing all targets (including any new ones) build.

**Documentation**

- [x] Documentation required for this feature
   * We need to add documentation and instruction for creating/managing WASM plugins. 

Doc PR is here: https://github.com/fluent/fluent-bit-docs/pull/881

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: .github/workflows/unit-tests.yaml [hunk 1]
```diff
@@ -50,7 +50,7 @@ jobs:
       - name: Setup environment
         run: |
           sudo apt update
-          sudo apt install -yyq gcc-7 g++-7 clang-6.0 libsystemd-dev gcovr libyaml-dev
+          sudo apt install -yyq gcc-7 g++-7 clang-6.0 libsystemd-dev gcovr libyaml-dev libedit-dev
```
Comment (patrick-stephens) [inline .github/workflows/unit-tests.yaml:53]: Do we need this dependency in the main builds under `packaging`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -110,6 +116,8 @@ option(FLB_AVRO_ENCODER        "Build with Avro encoding support" No)
 option(FLB_AWS_ERROR_REPORTER  "Build with aws error reporting support" No)
 option(FLB_ARROW               "Build with Apache Arrow support"  No)
 option(FLB_WINDOWS_DEFAULTS    "Build with predefined Windows settings"  Yes)
+option(FLB_WASM                "Build with WASM runtime support"  Yes)
```
Comment (patrick-stephens) [inline CMakeLists.txt:119]: For experimental stuff should we default to off?

File: CMakeLists.txt [hunk 3]
```diff
@@ -526,6 +536,29 @@ if(FLB_METRICS)
...
+       int main() {
+          _Atomic int a;
+          return 0;
+       }" FLB_HAVE_STDATOMIC_H)
+    check_include_files(stdatomic.h FLB_HAVE_STDATOMIC_H)
+    if (FLB_HAVE_STDATOMIC_H)
+      enable_language (ASM)
+      FLB_DEFINITION(FLB_HAVE_WASM)
+    else ()
+      message(STATUS "This platform is too old to support FLB_WASM feature. Now disabled.")
```
Comment (patrick-stephens) [inline CMakeLists.txt:553]: `too old` is probably not right, maybe just have 
```suggestion
      message(STATUS "This platform does not support the FLB_WASM feature so disabled.")
```

Implicit modification of flags though is probably not right: if someone explicitly enables FLB_WASM then I'd expect a build failure if I have missed dependencies or other set up.

**Review by patrick-stephens — COMMENTED** [review]
Only looked at the workflow and cmake config

File: CMakeLists.txt (hunk 2 above)
Comment (cosmo0920) [reply CMakeLists.txt:119]: Yeah, this should be default off but now turned on for CI.
Or, maybe we have to add `-DFLB_WASM=On` task on CI?

File: CMakeLists.txt (hunk 3 above)
Comment (cosmo0920) [reply CMakeLists.txt:553]: This mainly aimed for CentOS 7 and perhaps CentOS 6.
Enabling WASM feature is harmful to build for these platforms.
For removing this operation, we have to install devtoolset that provides stdatomic.h.

File: .github/workflows/unit-tests.yaml (hunk 1 above)
Comment (cosmo0920) [reply .github/workflows/unit-tests.yaml:53]: Ah, this line tries to fill dependencies for flb-wamrc that is AOT compiler using LLVM technology. Now, this should be disabled and no need to add. Should I revert it?

Comment (cosmo0920) [reply .github/workflows/unit-tests.yaml]: I removed it from the workflow fr now.

File: CMakeLists.txt (hunk 3 above)
Comment (cosmo0920) [reply CMakeLists.txt]: Fixed in https://github.com/fluent/fluent-bit/pull/5931/commits/63b9a285e761da6e41a77e815826344dc95252aa.

File: CMakeLists.txt [hunk 4]
```diff
@@ -110,6 +116,8 @@ option(FLB_AVRO_ENCODER        "Build with Avro encoding support"            No)
 option(FLB_AWS_ERROR_REPORTER  "Build with aws error reporting support"      No)
 option(FLB_ARROW               "Build with Apache Arrow support"             No)
 option(FLB_WINDOWS_DEFAULTS    "Build with predefined Windows settings"     Yes)
+option(FLB_WASM                "Build with WASM runtime support"            Yes)
```
Comment (patrick-stephens) [inline CMakeLists.txt:119]: Should we default to No?

**Review by patrick-stephens — COMMENTED** [review]
I think we should default to having this disabled personally.
If we do enable it then we need to ensure it works on all supported platforms as well. At the very least I'd like to see a full build of all targets to ensure there are no missing dependencies: https://github.com/fluent/fluent-bit/blob/master/packaging/local-build-all.sh

**Pull request merged**